source/s3_buffer_pool.c

```diff
@@ -0,0 +1,319 @@
+        return NULL;
+    }
+
+    struct aws_s3_buffer_pool *buffer_pool = aws_mem_calloc(allocator, 1, sizeof(struct aws_s3_buffer_pool));
+
+    AWS_FATAL_ASSERT(buffer_pool != NULL);
+
+    buffer_pool->base_allocator = allocator;
+    buffer_pool->block_size = block_size;
+    /* Somewhat arbitrary number. 
+     * Tries to balance between how many allocations use buffer and buffer space
+     * being wasted. */
+    buffer_pool->primary_size_cutoff = block_size / 4;
+    buffer_pool->mem_limit = mem_limit - s_buffer_pool_reserved_mem;
+    if (aws_mutex_init(&buffer_pool->mutex)) {
+        goto on_error;
+    }
+
+    aws_array_list_init_dynamic(&buffer_pool->blocks, allocator, s_min_num_blocks, sizeof(struct s3_buffer_pool_block));
+
+    return buffer_pool;
+
+on_error:
+    aws_mem_release(allocator, buffer_pool);
+    return NULL;
+}
+
+void aws_s3_buffer_pool_destroy(struct aws_s3_buffer_pool *buffer_pool) {
+    if (buffer_pool == NULL) {
+        return;
+    }
+
+    for (size_t i = 0; i < aws_array_list_length(&buffer_pool->blocks); ++i) {
+        struct s3_buffer_pool_block *block;
+        aws_array_list_get_at_ptr(&buffer_pool->blocks, (void **)&block, i);
+
+        AWS_FATAL_ASSERT(block->alloc_count == 0 && "Allocator still has outstanding blocks");
+        aws_mem_release(buffer_pool->base_allocator, block->block_ptr);
+    }
+
+    aws_array_list_clean_up(&buffer_pool->blocks);
+
+    aws_mutex_clean_up(&buffer_pool->mutex);
+    struct aws_allocator *base = buffer_pool->base_allocator;
+    aws_mem_release(base, buffer_pool);
+}
+
+void s_buffer_pool_trim_synced(struct aws_s3_buffer_pool *buffer_pool) {
+    for (size_t i = 0; i < aws_array_list_length(&buffer_pool->blocks); ++i) {
+        struct s3_buffer_pool_block *block;
+        aws_array_list_get_at_ptr(&buffer_pool->blocks, (void **)&block, i);
+
+        if (block->alloc_count == 0) {
+            aws_mem_release(buffer_pool->base_allocator, block->block_ptr);
+            aws_array_list_erase(&buffer_pool->blocks, i);
+            --i;
+        }
+    }
+}
+
+void aws_s3_buffer_pool_trim(struct aws_s3_buffer_pool *buffer_pool) {
+    aws_mutex_lock(&buffer_pool->mutex);
+    s_buffer_pool_trim_synced(buffer_pool);
+    aws_mutex_unlock(&buffer_pool->mutex);
+}
+
+struct aws_s3_buffer_pool_ticket *aws_s3_buffer_pool_reserve(struct aws_s3_buffer_pool *buffer_pool, size_t size) {
+
+    struct aws_s3_buffer_pool_ticket *ticket = NULL;
+    aws_mutex_lock(&buffer_pool->mutex);
+
+    size_t overall_taken = buffer_pool->primary_used + buffer_pool->primary_reserved + 
+                            buffer_pool->secondary_used + buffer_pool->secondary_reserved;
+
+    /*
+     * 
+     */
+    if (size > buffer_pool->primary_size_cutoff && 
+        (size + overall_taken) > buffer_pool->mem_limit &&
+        (buffer_pool->primary_allocated > (buffer_pool->primary_used + buffer_pool->primary_reserved + buffer_pool->block_size))) {
```

Comment:
trivial: yes, please add comment summarizing the logic, like "if this would exceed the limit, and a trim may help, then trim and see if we get any more memory"